### ISSUE TITLE
Optionally use iterative linear solvers for IDA when simulating very large systems

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#10815]
The data in the [ScalableTestSuite_NB](https://libraries.openmodelica.org/branches/heavy_tests/ScalableTestSuite_OB/ScalableTestSuite_OB.html) report allow to check how the time spent by the implicit solver of IDA to compute one step (excluding the time to compute residuals and Jacobians) scales with the system size.

Some examples scale very favourably. For example, the times taken by IDA to compute one step in CascadedFirstOrder

| Model                    |       Sim. time (s) | # of steps | time per step (ms) | 
| :----------------------- | -----------------:| --------------: | -----------: |
| CascadedFirstOrder_N_6400   |  0.80 | 2511 |         0.32 |
| CascadedFirstOrder_N_12800   |  2.44 | 3979 |         0.61 |
| CascadedFirstOrder_N_25600   |  7.43 | 6106|         1.22 |

scales as O(N), which is as favourable as it can get - in fact, it is the same scaling one would get with an explicit solver, without the stability limitations. The number of time steps increases with the problem size, but that is inherent in the dynamics of the system and the solver can't do much about it.

Other cases are much less favourable. For example, the time taken by IDA to compute a time step for HarmonicOscillator

| Model                    |       Sim. time (s) | # of steps | time per step (ms) | 
| :----------------------- | -----------------:| --------------: | -----------: |
| HarmonicOscillator_N_800   |  0.34 | 611 |       0.56  |
| HarmonicOscillator_N_1600   |  1.01 | 657|         1.54 |
| HarmonicOscillator_N_3200   |  5.65 | 569|         9.93 |

grows much faster than the system size. Currently, the build time still dominates for the range of sizes that the old backend can handle within a reasonable amount of time. However, as the new backend gets developed, the build time will be greatly reduced and only grow as O(N) because of the scalarized causalization phase, so the simulation time will become the bottleneck for sizes about one million and more variables, and quite badly so. In fact, the quadratic growth means that going above a certain model size becomes impractical.

I understand this heavily depends on the mathematical structure of the different problems. The complexity of solving a dense linear system is O(N^3), but for sparse systems, the exponent may actually depend on the system structure. For example, in CascadedFirstOrder, each state variable only influences the ones downstream of it, so the sparsity pattern is somewhat triangular, which justifies the favourable scaling. In Harmonic Oscillator, each state variable influences both the downstream and the upstream state derivatives, which probably leads to some kind of quadratic scaling property.

Bottom line: as the new backend allows us to tackle much larger systems beyond one million equations, the linear solver that we have been using so far for IDA (i.e., KLU) may no longer be adequate to the task.

According to the [Sundials documentation](https://sundials.readthedocs.io/en/latest/ida/Mathematics_link.html#nonlinear-solve), IDA supports both direct and iterative (Krylov) linear solvers out of the box. I think it's time we start considering adding this feature to the runtime.

In this case, we need to figure out how to handle the preconditioning matrix, which is a non-trivial topic.
